go/extractor/util/registryproxy.go

```diff
@@ -0,0 +1,140 @@
+package util
+
+import (
+	"encoding/json"
+	"fmt"
+	"log/slog"
+	"os"
+	"os/exec"
+	"strings"
+)
+
+const PROXY_HOST = "CODEQL_PROXY_HOST"
+const PROXY_PORT = "CODEQL_PROXY_PORT"
+const PROXY_CA_CERTIFICATE = "CODEQL_PROXY_CA_CERTIFICATE"
+const PROXY_URLS = "CODEQL_PROXY_URLS"
+const GOPROXY_SERVER = "goproxy_server"
+
+type RegistryConfig struct {
+	Type string `json:"type"`
+	URL  string `json:"url"`
+}
+
+// The address of the proxy including protocol and port (e.g. http://localhost:1234)
+var proxy_address string
+
+// The path to the temporary file that stores the proxy certificate, if any.
+var proxy_cert_file string
+
+// An array of registry configurations that are relevant to Go.
+// This excludes other registry configurations that may be available, but are not relevant to Go.
+var proxy_configs []RegistryConfig
+
+// Stores the environment variables that we wish to pass on to `go` commands.
+var proxy_vars []string = nil
+
+// Keeps track of whether we have inspected the proxy environment variables.
+// Needed since `proxy_vars` may be nil either way.
+var proxy_vars_checked bool = false
+
+// Tries to parse the given string value into an array of RegistryConfig values.
+func parseRegistryConfigs(str string) ([]RegistryConfig, error) {
+	var configs []RegistryConfig
+	if err := json.Unmarshal([]byte(str), &configs); err != nil {
+		return nil, err
+	}
+
+	return configs, nil
+}
+
+func getEnvVars() []string {
+	// If `proxy_vars` has been initialised, then we have already performed
+	// these checks and don't need to do so again. We assume that the environment
+	// variables are constant throughout the run of the autobuilder.
+	if proxy_vars != nil {
+		return proxy_vars
+	}
```

Comment:
It seems like overkill to do this check here and a very similar one at the only call site. I think it makes more sense to remove this one - then this function doesn't reference `proxy_vars` at all.